### ISSUE TITLE
Delete product will break the shopfront (during the current OC) of all users who have bought that item in the current OC

### ISSUE THREAD

**Issue by RachL** [openfoodfoundation/openfoodnetwork#3907]
## Description
When I log in with my producer account in Katuma's production I get a weird error and cannot do anything else.

## Steps to Reproduce
Just log in with my account rachel.arnould+carrots@gmail.com

Yesterday I was testing stuff and when I was done I turned my enterprise as "not visible" anymore as I didn't want any real customer to see it. Maybe this is having an impact?

## Animated Gif/Screenshot
![image](https://user-images.githubusercontent.com/6525576/59101584-b7b0ea00-8929-11e9-9e71-886b1d6ab360.png)

## Severity
bug-s2: a non-critical feature is broken, no workaround

## Your Environment
* Version used: v.2.0.2
* Browser name and version: Firefox and Chromium
* Operating System and version (desktop or mobile): Desktop and mobile

**Comment by luisramos0** [comment]
can you reproduce @RachL ? can you make the shop visible again and see if the error remains? would be good to know if it is related to that.

the error message points to a problem with the line items in the cart.
for reference, the code line where we are assigning line_item:
https://github.com/openfoodfoundation/openfoodnetwork/blob/568e3003ba43c713ac4fb5a57fddfbca086f7b98/app/assets/javascripts/darkswarm/services/cart.js.coffee#L13

**Comment by RachL** [comment]
@luisramos0 yes I can reproduce @sauloperez as well. But the thing is I cannot do anything on the platform logged in with that user. A super admin needs to make the shop visible again. I can't.

**Comment by sauloperez** [comment]
Good @luisramos0 I didn't find that skimming through the code. We need to investigate in DB what's wrong with that order and see which one is its root cause.

**Comment by luisramos0** [comment]
the order will probably be the cart order, the only latest order of the user.

**Comment by luisramos0** [comment]
Got it. Nice one!
@RachL you have bought Carrots and then deleted Carrots. That's causing this issue in the cart when showing bought items.

If you delete a product that was already sold as part of the current Order Cycle. You will break the cart and therefore the full frontoffice of all users who have bought that item in the current Order Cycle.

To replicate the issue:
- Add product to OC
- Checkout an order with that product
- Delete the product (while the OC is still open)
- Go back to shopfront

**Comment by luisramos0** [comment]
it's very unlikely a user deletes a product that was just sold. Can we downgrade this to S3?

In terms of fix I think we will need similar magic to [this](https://github.com/openfoodfoundation/spree/pull/41/files#diff-ba13136dc18703884af4dc9585973124R55) in the line_item.variant.product models.

**Comment by sauloperez** [comment]
Agree on both: better as an S3 and it immediately sounded like a soft-delete issue.

**Comment by luisramos0** [comment]
This is hard, I am not taking it now...

Even if we handle line_item.variant.product, for example, the current order serializer is loading the variants in the order which include variant.price which is not present in the DB after the product is deleted... there are many approaches to this, none is a few minutes job:
- remove variant serialization from current_order serialization
- make spree::price soft deletable
- make variant code handle the fact the price may not be there...

**Comment by luisramos0** [comment]
Another option to this issue is to prevent the user from deleting a product that is in an open order cycle (or was sold in one and then removed). We would need to prevent OCs with deleted products to be re-opened.

**Comment by sauloperez** [comment]
I haven't checked the code but the most reasonable solution seems to be

> make spree::price soft deletable 

It might make sense to delete a product from an OC in some situations.

**Comment by sauloperez** [comment]
This one is indeed very annoying, even in development

![Screenshot from 2019-06-18 18-12-49](https://user-images.githubusercontent.com/762088/59700930-bd91af80-91f4-11e9-89ed-8350cd13f95b.png)

**Comment by lin-d-hop** [comment]
An issue related to this has popped up on UK production. The producer deleted a product that was sold in an OC. The OC was closed so fortunately we didn't get broken shopfronts. However the hub manager now can't delete the product from invoices etc. It is a different issue but related.

From my understanding of the Network Feature conversations, we've spoken of never being able to delete a product. Other major actors in this space follow that approach. How we handle product delete is also important for Delivery Note and legal Invoicing. 

Maybe I am not understanding fully. Forgive me if so. But maybe soft-delete is too much. Maybe we need to be thinking about archiving products?

**Comment by luisramos0** [comment]
soft delete will do the job. its not actaully a delete. it just marks the
product with a timestamp  on a column "deleted at".
i think you should open a separate issue and specify exactly what that
manager needs to do in invoices. ok?

**Comment by sigmundpetersen** [comment]
Ok I put this as test ready blocked by merge of #4204
Same as #4203, we can test after merge and see if it's fixed.

Please move it back to dev if I'm wrong @Matt-Yorkley

**Comment by RachL** [comment]
I can't reproduce. Closing :tada:
